spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java

```diff
@@ -133,23 +139,34 @@ private void updateNotPropagatedHeaders(String[] headers, boolean merge) {
 			}
 			this.notPropagatedHeaders.addAll(Arrays.asList(headers));
 		}
+		boolean hasAsterisk = this.notPropagatedHeaders.contains("*");
+
+		if (hasAsterisk) {
+			this.notPropagatedHeaders.clear();
+			this.notPropagatedHeaders.add("*");
+		}
```

Comment:
I would just set a boolean in this case so we can optimize later e.g `this.copyNoHeaders=true`.

Reply:
I do that. See below:
```
this.shouldCopyRequestHeaders = !hasAsterisk;
```
And `shouldCopyRequestHeaders()` body.
What am I missing?

Reply:
Ah - didn't see that but, since subclasses can (and do) override that method, we need another boolean.

Reply:
Right, and they always override it to `false` eliminating all the purpose of this `notPropagatedHeaders`

Reply:
While ours might override to `false`, we can't guarantee user handlers do the same.